src/main/java/com/github/stefvanschie/inventoryframework/GuiListener.java

```diff
@@ -98,9 +128,51 @@ public void onInventoryClose(@NotNull InventoryCloseEvent event) {
             onClose.accept(event);
         }
 
-        HumanEntityCache humanEntityCache = gui.getHumanEntityCache();
-        HumanEntity humanEntity = event.getPlayer();
-        humanEntityCache.restore(humanEntity);
-        humanEntityCache.clearCache(humanEntity);
+        gui.getHumanEntityCache().restoreAndForget(event.getPlayer());
+
+        if (gui.getViewerCount() == 1) {
+            activeGuiInstances.remove(gui);
+        }
+    }
+
+    /**
+     * Registers newly opened inventories
+     *
+     * @param event the event fired
+     * @since 0.5.19
+     */
+    @EventHandler(ignoreCancelled = true)
+    public void onInventoryOpen(@NotNull InventoryOpenEvent event) {
+        if (!(event.getInventory().getHolder() instanceof Gui)) {
+            return;
+        }
+
+        Gui gui = (Gui) event.getInventory().getHolder();
+        activeGuiInstances.add(gui);
+    }
+
+    /**
+     * Handles the disabling of the plugin
+     *
+     * @param event the event fired
+     * @since 0.5.19
+     */
+    @EventHandler(priority = EventPriority.HIGHEST, ignoreCancelled = true)
+    public void onPluginDisable(@NotNull PluginDisableEvent event) {
+        if (event.getPlugin() != plugin) {
+            return;
+        }
+
+        int counter = 0; //callbacks might open GUIs, eg. in nested menus
+        while (!activeGuiInstances.isEmpty() && counter++ < 10) {
+            for (Gui gui : new ArrayList<>(activeGuiInstances)) {
+                for (HumanEntity viewer : gui.getViewers()) {
+                    viewer.closeInventory();
+                }
+            }
+        }
+
+        plugin.getLogger().warning("Unable to close GUIs on plugin disable: they keep getting opened "
+                + "(tried: " + counter + " times)");
```

Comment:
This message is always sent even if the gui was able to close within one attempt and no other guis were opened. This should probably only be sent if the amount of trials has been exceeded (in this case 10) and there are still open guis left.